homeassistant/components/recorder/run_history.py

```diff
@@ -0,0 +1,144 @@
+
+    def get(self, start: datetime) -> RecorderRuns | None:
+        """Return the recorder run that started before or at start.
+
+        If the first run started after the start, return None
+
+        This function is thread safe.
+        """
+        if start >= self.recording_start:
+            return self.current
+
+        # Hold a reference here to self._run_history
+        # to ensure it does not get changed out from
+        # under us in the recorder thread.
+        run_history = self._run_history
+
+        # Ensure we do not access self._run_history
+        # after this point as the other thread
+        # could change it out on us
+        run_timestamps = run_history.run_timestamps
+        runs_by_timestamp = run_history.runs_by_timestamp
+
+        # bisect_left tells us were we would insert
+        # a value in the list of runs after the start timestamp.
+        #
+        # The run before that (idx-1) is when the run started
+        #
+        # If idx is 0, history never ran before the start timestamp
+        #
+        if idx := bisect.bisect_left(run_timestamps, start.timestamp()):
+            return runs_by_timestamp[run_timestamps[idx - 1]]
+        return None
+
+    def start(self, session: Session) -> None:
+        """Start a new run.
+
+        Must run in the recorder thread.
+        """
+        self._current_run_info = RecorderRuns(
+            start=self.recording_start, created=dt_util.utcnow()
+        )
+        session.add(self._current_run_info)
+        session.flush()
+        session.expunge(self._current_run_info)
+        self.load_from_db(session)
+
+    def reset(self) -> None:
+        """Reset the run when the database is changed or fails.
+
+        Must run in the recorder thread.
+        """
+        self._recording_start = dt_util.utcnow()
+        self._current_run_info = None
+
+    def end(self, session: Session) -> None:
+        """End the current run.
+
+        Must run in the recorder thread.
+        """
+        assert self._current_run_info is not None
+        self._current_run_info.end = dt_util.utcnow()
+        session.add(self._current_run_info)
+
+    def load_from_db(self, session: Session) -> None:
+        """Update the run cache.
+
+        Must run in the recorder thread.
+        """
+        run_timestamps: list[int] = []
+        runs_by_timestamp: dict[int, RecorderRuns] = {}
+
+        for run in session.query(RecorderRuns).order_by(RecorderRuns.start.asc()).all():
+            session.expunge(run)
+            if run_dt := process_timestamp(run.start):
+                timestamp = run_dt.timestamp()
+                run_timestamps.append(timestamp)
+                runs_by_timestamp[timestamp] = run
+
+        #
+        # self._run_history is accessed in get()
+        # which is allowed to be called from any thread
+        #
+        # We use a dataclass to ensure that when we update
+        # self._run_history it is done so atomiclly to avoid
```

Comment:
How do dataclasses guarantee atomicity?

Reply:
Bad choice of words here since what I'm trying to say is that both values depend on each other and since they are in the dataclass one can't be updated without the other.  Will rephrase.